src/docs/design-patterns/decorator-pattern.md

```diff
@@ -0,0 +1,214 @@
+
+- **Extension Without Modification**: In situations where modifying the original class is not possible due to restrictions like access levels or when the class is part of a library you do not control, decorators provide a way to add behavior.
+
+- **Conditional Behavior Addition**: When the behavior needs to be added to objects only under certain conditions or contexts, decorators can be dynamically applied and removed.
+
+- **Enhancing Legacy Code**: For enhancing or fixing legacy code where changing the original codebase is risky or impractical, applying decorators can introduce new behavior or fix issues without touching the original code.
+
+- **Compliance and Logging**: In scenarios requiring compliance, logging, or audit trails where these concerns are secondary to the main business logic, decorators can inject these cross-cutting concerns transparently.
+
+- **Validation and Security**: To add validation, authentication, or other security layers around core functionalities without embedding these concerns directly into the primary classes.
+
+- **Flexible Alternative to Inheritance**: When inheritance is too rigid or not suitable for adding functionalities, the Decorator pattern offers a more flexible solution by allowing mixing and matching of behaviors.
+
+Utilizing the Decorator pattern under these circumstances can lead to cleaner, more maintainable, and scalable code architectures by promoting separation of concerns and adherence to the Open/Closed Principle.
+
+## C# Example
+
+Let's implement the Decorator Pattern in a .NET application. We'll start with a basic `BookService` that adds a book to a repository. Then, we'll add decorators for logging with a stopwatch/timer and for validating the book using [FluentValidation](https://docs.fluentvalidation.net/en/latest/).
+
+### The Underlying Service
+
+First, define the `IBookService` interface and implement a basic `BookService`.
+
+```csharp
+public interface IBookService
+{
+    void AddBook(Book book);
+}
+
+public class BookService : IBookService
+{
+    private readonly IRepository<Book> _repository;
+
+    public BookService(IRepository<Book> repository)
+    {
+        _repository = repository;
+    }
+
+    public void AddBook(Book book)
+    {
+        _repository.Add(book);
+    }
+}
+```
+
+### Implement a Logging Decorator
+
+The logging decorator will log how long it takes to add a book to the repository:
+
+```csharp
+public class LoggingBookServiceDecorator : IBookService
+{
+    private readonly IBookService _decoratedBookService;
+    private readonly ILogger<LoggingBookServiceDecorator> _logger;
+
+    public LoggingBookServiceDecorator(IBookService decoratedBookService, ILogger<LoggingBookServiceDecorator> logger)
+    {
+        _decoratedBookService = decoratedBookService;
+        _logger = logger;
+    }
+
+    public void AddBook(Book book)
+    {
+        _logger.LogInformation($"Adding a book: {book}", book);
+        var stopwatch = Stopwatch.StartNew();
```

Comment:
With the call to Stopwatch, you need to include `using System.Diagnostics;` at the top. Calling out `using` statements is a good practice.